sdk/python/kfp/v2/compiler/compiler.py

```diff
@@ -0,0 +1,317 @@
+  static and dynamic loops, etc.
+
+  Example:
+    How to use the compiler to construct pipeline_spec json:
+
+      @dsl.pipeline(
+        name='name',
+        description='description'
+      )
+      def my_pipeline(a: int = 1, b: str = "default value"):
+        ...
+
+      Compiler().compile(my_pipeline, 'path/to/pipeline.json')
+  """
+
+  def _create_pipeline_spec(
+      self,
+      args: List[dsl.PipelineParam],
+      pipeline: dsl.Pipeline,
+  ) -> pipeline_spec_pb2.PipelineSpec:
+    """Create the pipeline spec object.
+
+    Args:
+      args: The list of pipeline arguments.
+      pipeline: The instantiated pipeline object.
+
+    Returns:
+      The IR proto of pipeline_spec.
+
+    Raises:
+      NotImplementedError if the argument is of unsupported types.
+    """
+    pipeline_spec = pipeline_spec_pb2.PipelineSpec()
+    pipeline_spec.pipeline_info.name = pipeline.name or 'Pipeline'
+    pipeline_spec.sdk_version = 'kfp.v2'
+    pipeline_spec.schema_version = 'dummy.v1'
+
+    # Pipeline Parameters
+    for arg in args:
+      if arg.value is not None:
+        if isinstance(arg.value, int):
+          pipeline_spec.runtime_parameters[
+              arg.name].type = pipeline_spec_pb2.PrimitiveType.INT
+          pipeline_spec.runtime_parameters[
+              arg.name].default_value.int_value = arg.value
+        elif isinstance(arg.value, float):
+          pipeline_spec.runtime_parameters[
+              arg.name].type = pipeline_spec_pb2.PrimitiveType.DOUBLE
+          pipeline_spec.runtime_parameters[
+              arg.name].default_value.double_value = arg.value
+        elif isinstance(arg.value, str):
+          pipeline_spec.runtime_parameters[
+              arg.name].type = pipeline_spec_pb2.PrimitiveType.STRING
+          pipeline_spec.runtime_parameters[
+              arg.name].default_value.string_value = arg.value
+        else:
+          raise NotImplementedError(
+              'Unexpected parameter type with: "{}".'.format(str(arg.value)))
+
+    deployment_config = pipeline_spec_pb2.PipelineDeploymentConfig()
+    importer_tasks = []
+
+    def _get_input_artifact_type(
+        input_name: str,
+        component_spec: structures.ComponentSpec,
+    ) -> str:
+      """Find the input artifact type by input name.
+
+      Args:
+        input_name: The name of the component input.
+        component_spec: The component spec object.
+
+      Returns:
+        The input type name if found in component_spec, or '' if not found.
+      """
+      for component_input in component_spec.inputs:
+        if component_input.name == input_name:
+          return type_utils.get_artifact_type(component_input.type)
+      return ''
+
+    for op in pipeline.ops.values():
+      component_spec = op._metadata
+      task = pipeline_spec.tasks.add()
+      task.CopyFrom(op.task_spec)
+      deployment_config.executors[task.task_info.name].container.CopyFrom(
```

Comment:
qq: Do we need to dedup the executor label here in case that one component spec is imported twice to generate two tasks?

Reply:
> Could there be a legit use case of importing the same spec twice?

For example, parallel training using different spec or hyperparam. This was the actual use case by one of our users trying to experiment and compare between multiple models.

> Maybe I need to have some logic to append a sequence number there?

Likely yes. One constraint imposed by IR is that the task names are unique within the same pipeline. We can safely align the executor label with task name then.